### ISSUE TITLE
Incorrect result may caused by optimization

### ISSUE THREAD

**Issue by Syang111** [cockroachdb/cockroach#120049]
**Describe the problem**
In theory, sql1 and sql2 should have the same result. However, sql1 return 0 while sql2 return nothing. It seems a logical bug.
```
SELECT t0.c0 FROM t0 WHERE (t0.c0 IN ((1805922729-1791620580) * BIT_LENGTH ('') , 547073614, NULL )) IS TRUE;  --sql1
  c0
------
   0
(1 row)

SELECT c0 FROM (SELECT c0, ( t0.c0  IN ((1805922729-1791620580 )* BIT_LENGTH ('') , 547073614, NULL )) IS TRUE AS flag FROM t0) AS t WHERE flag; --sql2
  c0
------
(0 rows)

```
**To Reproduce**

1. I run cockroach with command "sudo docker run -d --name cockroach -p 26257:26257 cockroachdb/cockroach:latest-v23.2 start-single-node --insecure
2. I executed command "./cockroach sql --insecure" to interact with cockroach
3. I execute following statements in CLI command, and got an unexpected result.

```
CREATE DATABASE database0;
USE database0;
CREATE TABLE t0 (c0 INT2 DEFAULT (((738228170)/(((((true) :::BOOL)) ::INT2)))), CONSTRAINT "primary" PRIMARY KEY(c0 ASC), FAMILY "primary" (c0));
CREATE TABLE t1 (c0 FLOAT AS (0.9950516636813965) STORED, c1 VARBIT(51) AS (B'1111111111111111111111111111111110001111001101110110001011111010') STORED NOT NULL , FAMILY "primary" (c0, c1));
UPSERT INTO t0 (c0) VALUES(0);

root@localhost:26257/database4> SELECT t0.c0 FROM t0 WHERE (t0.c0 IN ((1805922729-1791620580) * BIT_LENGTH ('') , 547073614, NULL )) IS TRUE;                                                               
  c0
------
   0
(1 row)

root@localhost:26257/database4> SELECT c0 FROM (SELECT c0, ( t0.c0  IN ((1805922729-1791620580 )* BIT_LENGTH ('') , 547073614, NULL )) IS TRUE IS TRUE AS flag FROM t0) AS t WHERE flag;                    
  c0
------
(0 rows)
```

**Expected behavior**
Both query should return the same result.

**Environment:**
 - Client version: CockroachDB CCL v23.1.0-beta.2 (x86_64-pc-linux-gnu, built 2023/04/13 14:55:11, go1.19.4)
 - Server version: CockroachDB CCL v23.2.0 (x86_64-pc-linux-gnu, built 2024/01/16 19:30:18, go1.21.5 X:nocoverageredesign)
 - Server OS: Linux
 - Client app: `cockroach sql`

Jira issue: CRDB-36449

**Comment by DrewKimball** [comment]
Hi @Syang111, thanks for the report! It looks like there's a bug in the vectorized engine:
```
root@localhost:26257/database0> select c0, c0 in ((1805922729-1791620580 )* BIT_LENGTH ('') , 547073614, NULL) from t0;
  c0 | ?column?
-----+-----------
   0 |   NULL
(1 row)

Time: 5ms total (execution 4ms / network 1ms)

root@localhost:26257/database0> set vectorize=off;
SET

Time: 2ms total (execution 1ms / network 0ms)

root@localhost:26257/database0> select c0, c0 in ((1805922729-1791620580 )* BIT_LENGTH ('') , 547073614, NULL) from t0;
  c0 | ?column?
-----+-----------
   0 |    t
(1 row)

Time: 4ms total (execution 4ms / network 0ms)
```

**Comment by DrewKimball** [comment]
TL;DR is that we use a binary search to implement the `IN` operator, and the `547073614` value overflows `int16` and causes the tuple to become unsorted. I'm going to close this issue as a dupe of #102864, but feel free to ask any further questions!
